compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmObservabilityGen.java

```diff
@@ -222,66 +231,88 @@ void instrumentPackage(BIRPackage pkg) {
     /**
      * Adding Java Interop calls to basic blocks.
      * Here the JI calls are added for all kinds of terminators.
-     *
      * First we check if there are position details for instructions, if present we add the JI calls with those
      * positions else, we consider the terminator position to create the JI call.
      *
      * @param func The function of which the instructions should be rewritten
      * @param pkg The package containing the function
      */
     private void rewriteControlFlowInvocation(BIRFunction func, BIRPackage pkg) {
-        int i = 0;
-        while (i < func.basicBlocks.size()) {
-            // Basic blocks with JI method calls are added for all kinds of Terminators
-            BIRBasicBlock currentBB = func.basicBlocks.get(i);
-
-            // First we give the priority to Instructions,
-            // If no instructions are found, then we get the Terminator position
-            Location desugaredPos = currentBB.terminator.pos;
-            for (BIRNonTerminator instruction : currentBB.instructions) {
-                if (instruction.pos != null) {
-                    desugaredPos = instruction.pos;
-                    break;
-                }
-            }
+        populatePredecessorMap(func.basicBlocks);
+        for (Map.Entry<BIRBasicBlock, List<BIRBasicBlock>> entry : this.predecessorMap.entrySet()) {
+            BIRBasicBlock currentBB = entry.getKey();
+            Location desugaredPos = getDesugaredPosition(currentBB);
 
             if (desugaredPos != null && desugaredPos.lineRange().startLine().line() >= 0) {
-                BIRBasicBlock newBB = insertBasicBlock(func, i + 1);
-                swapBasicBlockContent(func, currentBB, newBB);
-                injectCheckpointCall(func, currentBB, pkg, desugaredPos);
-                currentBB.terminator.thenBB = newBB;
-                // Fix error entries in the error entry table
-                fixErrorTable(func, currentBB, newBB);
-                i += 1; // Number of inserted BBs
+                List<BIRBasicBlock> predecessors = entry.getValue();
+                boolean desugaredPosAlreadyLoaded = predecessors.stream()
```

Comment:
Shall we simplify this without using streams

Reply:
Since it's in compile time, would it affect the performance?